### PR TITLE
Rdp client refactor

### PR DESCRIPTION
**Pull request by ibeckermayer**
See the individual commits and their messages

### CODE REVIEW

File: lib/srv/desktop/rdp/rdpclient/src/lib.rs [hunk 1]
```diff
@@ -120,6 +121,32 @@ pub unsafe extern "C" fn client_run(cgo_handle: CgoHandle, params: CGOConnectPar
...
+        match $operation {
+            Ok(_) => CGOErrCode::ErrCodeSuccess,
+            Err(e) => {
+                error!(concat!($ctx, " failed: {:?}"), e);
+                CGOErrCode::ErrCodeFailure
+            }
+        }
+    };
+}
+
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/src/lib.rs:149]: I really don't like macros with flow control (like `return` below), can we replace them both with regular function?
For example:
```suggestion
fn handle_operation<T>(cgo_handle: CgoHandle, ctx: &'static str, f: T) -> CGOErrCode
    where
        T: FnOnce(ClientHandle) -> ClientResult<()>,
{
    let client_handle = match get_client_handle(cgo_handle) {
        Some(it) => it,
        None => {
            warn!("call_function_on_handle failed: handle not found");
            return CGOErrCode::ErrCodeFailure;
        }
    };
    match f(client_handle) {
        Ok(_) => CGOErrCode::ErrCodeSuccess,
        Err(e) => {
            error!("{} failed: {:?}", ctx, e);
            CGOErrCode::ErrCodeFailure
        }
    }
}
```

Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/lib.rs]: It took a little more finagling to be able to pass in arbitrary arguments:

b4ce750f40e777372a37c5144b4be7a177b5a6bb

Comment (probakowski) [reply lib/srv/desktop/rdp/rdpclient/src/lib.rs]: You don't have to handle variable arguments, compiler can turn closure into `FnOnce`,  with my proposal above call site can look like: 
```rust
Ok(s) => handle_operation(
            cgo_handle,
            "client_update_clipboard",
            |client_handle| client_handle.update_clipboard(s),
        ),
```

Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/lib.rs]: Oh yeah, my bad.

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs [hunk 2]
```diff
@@ -768,100 +753,116 @@ impl std::fmt::Display for FilesystemBackendError {
...
+    }
+
+    fn insert(&mut self, completion_id: CompletionId, handler: T) {
+        self.cache.insert(completion_id, handler);
+    }
+
+    fn remove(&mut self, completion_id: &CompletionId) -> Option<T> {
+        self.cache.remove(completion_id)
+    }
+}
```
Comment (ibeckermayer) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs]: Moving https://github.com/gravitational/teleport/pull/34011#discussion_r1376829352 to here.

Yes that's a better idea, see: c9913e7b1bde102a6649c98d818636df8a6db093

**Review by probakowski — APPROVED** [review]

**Pull request merged**